### DISCUSSION TITLE
How do obfs work?

### DISCUSSION THREAD

**Discussion by mariush444** [osmandapp/OsmAnd#12094] [Q&A] [answered]
Can anyone can explain to me how obfs work. please?
I'm just curious, I wonder. 

We can have some main maps:
World_basemap
World_basemap_mini
Country/Region e.g. Malta_europe
**How does osmand "decide" which map to show.
Is it because of zoom that is defined inside the obf?**
**When I create my own Malta_my_map_obf (additionally to Malta_europe) with the same zooms as Malta_europe
which one will be choos by osmand and why.**

SRTM:
We have obf for srtm
How osmand work with it? 
Is it something like overlay/underlay?
**How is rendered?
Does osmand show this map because of the name Malta_europe_srtm or try to find any map with proper coordinates?**

Live:
**Do this file overwrite/modify parts of the existing country map?
Is it another semi overlay/underlay?**

If someone find a time to explain this to newby it would be great.

**Comment by vshcherb** [comment]
Yes, zoom levels, inside. Please check inspector.sh utility to introspect the contentss fo OBF. This is incomplete article https://docs.osmand.net/en/main@latest/development/osmand-file-formats/osmand-obf, though there is a complete analogue of rendering https://docs.osmand.net/en/main@latest/development/osmand-file-formats/osmand-rendering-style

**Reply by mariush444** [reply]
thank you

**Comment by thompehw** [comment, marked as answer]
Here's something about live updates that I wrote on the old Google support group - it's just my amateur understanding:
>
> The live updates become offline incremental updates to the main map file once they are downloaded - hourly, daily, weekly, or on-demand.(the files go into a folder called "live"). Presumably OsmAnd servers make an overpass query to determine how OSM has changed since the previous update, then run this through OsmAndMapCreator to make the update .obf file. Each time you update you get another file saying what has changed since the previous update, it doesn't go all the way back to the map creation date each time.
>
> The update files have a date/time suffix to their filename such that the most recent updates are later in filename sort order. An object in a later file replaces an object having the same ID in an earlier file. Objects with new IDs are added. An object is tagged with osmand_change=delete if it is to be removed.
>
>I hope someone finds that interesting! You can even customise your map by making local edits to OSM objects in JOSM, saving as .osm, then using MapCreator to make a .obf update file with a suitable filename. You don't need to be subscribed to OsmAnd Live for this to work.

**Reply by mariush444** [reply]
thanks a lot.
I see that there are region_YY_MM_00.obf and region_YY_MM_DD.obf in live directory
I thought that  region_YY_MM_00.obf was something like base/initial state that was updated by region_YY_MM_DD.obf
But region_YY_MM_00.obf is updated at the same time as new region_YY_MM_DD.obf is downloaded.
Do you know/guess why?

Do you know/guess how it's rendered?
Let's assume we have straight street in main map region.obf
then we downloaded region_YY_MM_00.obf we the same street but corrected/changed and now the street goes to the left,
and next we have region_YY_MM_D.obf with info that this street has 2 lines and not one as was in main region.obf and region_YY_MM_00.obf .

What will we see?
1. straight street AND street goes to the left with one line AND street's going to the left with 2 lines
2. street's going  to the left with 2 lines only (straight street in the region.obf will be "erased")

**Reply by thompehw** [reply]
Assuming the ID of the street is not changed during these edits (i.e. it is not deleted in OSM and redrawn as a new object), the I think you would see it with the geometry and tagging given in the OBF with the last filename sort order - information about the same object in the earlier (and original) OBFs will be discarded.

(EDIT: Actually, even if the object were deleted and re-drawn then the result would be the same, because the original (deleted) object would appear in one of the updates with the Osmand-specific tag of osmand_change=delete, which prevents it from being rendered.)

By the way, the same system (objects in OBFs with a later filename sort order supersede earlier objects with the same ID) also works for OBFs in the base folder where OsmAnd puts its main map downloads - in this respect there's nothing special about the "live" folder except that if you are subscribed to Live that is where the update OBFs happen to be placed.

I don't know the answer to your earlier question about the role of the *_00.obf files when live updates are activated.

**Comment by thompehw** [comment]
On your SRTM questions, here's are some thoughts based on my experiments as a user:

Contour lines are precomputed vector data that work the same was as e.g. a highway in OSM, but with suitable tagging for OsmAnd to recognise what they are. They are rendered along with the map data under control of the map rendering style file *.render.xml

It's possible to generate your own contours using something like QGIS then OsmAndMapCreator.

Just as OSM nodes have coordinates, so do the nodes along a contour (at each point it changes direction), so the .obf filename is not controlling where they appear on the map - it just shows you which download you need for your region.

OsmAnd draws the contours without understanding anything about elevation from them. The elevation data to support elevation-aware routing must be contained in the main map OBF, and only seems to work for navigation along ways. Thus, if you drop a pin in the middle of a field and want to know its elevation, OsmAnd can't tell you (people have suggested this feature). You can work it out by looking at the contour lines yourself, or you can plot a route to it and look at the elevation of the point on a way that gets closest.

I assume that hillshades and slopes are raster images that are displayed like an underlay.

I look forward to being corrected if any of this is wrong! All the best.

**Reply by mariush444** [reply]
many many thanks for knowledge that you shared

**Comment by KOLANICH** [comment]
Could you, please, add into the FAQ the answer to the following question:

>What is the rationale for using the custom map format instead of OSM pbf with some additional index files?
